src/types/redis_hash.cc

```diff
@@ -418,4 +469,272 @@ rocksdb::Status Hash::RandField(const Slice &user_key, int64_t command_count, st
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisHash);
+  batch->PutLogData(log_data.Encode());
+
+  // expire special field
+  std::vector<rocksdb::PinnableSlice> values;
+  values.resize(sub_keys.size());
+  std::vector<rocksdb::Status> statuses;
+  statuses.resize(sub_keys.size());
+  storage_->MultiGet(read_options, storage_->GetDB()->DefaultColumnFamily(), keys.size(), keys.data(), values.data(),
+                     statuses.data());
+
+  auto now = util::GetTimeStampMS();
+  for (size_t i = 0; i < keys.size(); i++) {
+    if (!statuses[i].ok() && !statuses[i].IsNotFound()) return statuses[i];
+
+    // no such field exists
+    if (statuses[i].IsNotFound()) {
+      ret->emplace_back(-2);
+      continue;
+    }
+
+    // expire with a pass time
+    if (expire_ms <= now) {
+      batch->Delete(sub_keys[i]);
+      ret->emplace_back(2);
+      metadata.size -= 1;
+      continue;
+    }
+
+    auto value = values[i].ToString();
+    uint64_t field_expire = 0;
+    decodeExpireFromValue(metadata, &value, field_expire);
+
+    // if a field has no associated expiration, we treated it expiration is infinite
+    auto treated_expire = field_expire == 0 ? UINT64_MAX : field_expire;
+    if (type == HashFieldExpireType::None || (type == HashFieldExpireType::NX && field_expire == 0) ||
+        (type == HashFieldExpireType::XX && field_expire != 0) ||
+        (type == HashFieldExpireType::GT && expire_ms > treated_expire) ||
+        (type == HashFieldExpireType::LT && expire_ms < treated_expire)) {
+      encodeExpireToValue(&value, expire_ms);
+      batch->Put(sub_keys[i], value);
+      // 1 if expiration was updated
+      ret->emplace_back(1);
+    } else {
+      // 0 if condition has not been met
+      ret->emplace_back(0);
+    }
+  }
+
+  // convert rest field encoding
+  if (!metadata.IsFieldExpirationEnabled()) {
+    metadata.field_encoding = HashSubkeyEncoding::VALUE_WITH_TTL;
+
+    std::unordered_set<std::string_view> field_set;
+    for (auto field : fields) {
+      if (!field_set.emplace(field.ToStringView()).second) {
+        continue;
+      }
+    }
+
+    std::string prefix_key = InternalKey(ns_key, "", metadata.version, storage_->IsSlotIdEncoded()).Encode();
+    std::string next_version_prefix_key =
+        InternalKey(ns_key, "", metadata.version + 1, storage_->IsSlotIdEncoded()).Encode();
+
+    rocksdb::Slice upper_bound(next_version_prefix_key);
+    read_options.iterate_upper_bound = &upper_bound;
+
+    auto iter = util::UniqueIterator(storage_, read_options);
+    for (iter->Seek(prefix_key); iter->Valid() && iter->key().starts_with(prefix_key); iter->Next()) {
+      InternalKey sub_ikey(iter->key(), storage_->IsSlotIdEncoded());
+      auto value = iter->value().ToString();
+      if (field_set.find(sub_ikey.GetSubKey().ToStringView()) == field_set.end()) {
+        encodeExpireToValue(&value, 0);
+        batch->Put(sub_ikey.Encode(), value);
+      }
+    }
```

Comment:
It's not good to do this because it will be too slow if a hash has a lot of fields. Even having the old hash key not allow to expire fields is better than the current implementation. cc @PragmaTwice @torwig

Reply:
Maybe we can add a option in config file, that lets users decide whether to allow old hash to be converted to the new format?
If the user clearly knows that their hash is not large and wants to support field expiration on existing hash object. It is very useful.

Reply:
I do not prefer doing this since it will introduce an extra maintenance burden. It makes sense only to allow to expire fields in the new hash key from my side.

Reply:
Yeah, please don't perform the convertion automatically .

Reply:
also we should to provide an option for users to let them decide if the newly created hash object is in the new encoding.

Reply:
@git-hulk @PragmaTwice Hi, I have modified it. Don't encode field ttl on existing hash object. please review.